File: examples/go/data_structures.go
package main

import (
	"fmt"
	"math"
	"sort"
)

// Generic data structures

type Stack struct {
	items []interface{}
}

func NewStack() *Stack {
	return &Stack{items: make([]interface{}, 0)}
}

func (s *Stack) Push(item interface{}) {
	s.items = append(s.items, item)
}

func (s *Stack) Pop() (interface{}, error) {
	if len(s.items) == 0 {
		return nil, fmt.Errorf("stack is empty")
	}
	item := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return item, nil
}

func (s *Stack) Peek() (interface{}, error) {
	if len(s.items) == 0 {
		return nil, fmt.Errorf("stack is empty")
	}
	return s.items[len(s.items)-1], nil
}

func (s *Stack) Size() int {
	return len(s.items)
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}

type Queue struct {
	items []interface{}
}

func NewQueue() *Queue {
	return &Queue{items: make([]interface{}, 0)}
}

func (q *Queue) Enqueue(item interface{}) {
	q.items = append(q.items, item)
}

func (q *Queue) Dequeue() (interface{}, error) {
	if len(q.items) == 0 {
		return nil, fmt.Errorf("queue is empty")
	}
	item := q.items[0]
	q.items = q.items[1:]
	return item, nil
}

func (q *Queue) Front() (interface{}, error) {
	if len(q.items) == 0 {
		return nil, fmt.Errorf("queue is empty")
	}
	return q.items[0], nil
}

func (q *Queue) Size() int {
	return len(q.items)
}

func (q *Queue) IsEmpty() bool {
	return len(q.items) == 0
}

// Binary Tree implementation
type TreeNode struct {
	Value int
	Left  *TreeNode
	Right *TreeNode
}

type BinaryTree struct {
	Root *TreeNode
}

func NewBinaryTree() *BinaryTree {
	return &BinaryTree{Root: nil}
}

func (bt *BinaryTree) Insert(value int) {
	bt.Root = bt.insertNode(bt.Root, value)
}

func (bt *BinaryTree) insertNode(node *TreeNode, value int) *TreeNode {
	if node == nil {
		return &TreeNode{Value: value}
	}
	
	if value < node.Value {
		node.Left = bt.insertNode(node.Left, value)
	} else {
		node.Right = bt.insertNode(node.Right, value)
	}
	return node
}

func (bt *BinaryTree) Search(value int) bool {
	return bt.searchNode(bt.Root, value)
}

func (bt *BinaryTree) searchNode(node *TreeNode, value int) bool {
	if node == nil {
		return false
	}
	
	if value == node.Value {
		return true
	} else if value < node.Value {
		return bt.searchNode(node.Left, value)
	} else {
		return bt.searchNode(node.Right, value)
	}
}

func (bt *BinaryTree) InorderTraversal() []int {
	var result []int
	bt.inorder(bt.Root, &result)
	return result
}

func (bt *BinaryTree) inorder(node *TreeNode, result *[]int) {
	if node != nil {
		bt.inorder(node.Left, result)
		*result = append(*result, node.Value)
		bt.inorder(node.Right, result)
	}
}

func (bt *BinaryTree) PreorderTraversal() []int {
	var result []int
	bt.preorder(bt.Root, &result)
	return result
}

func (bt *BinaryTree) preorder(node *TreeNode, result *[]int) {
	if node != nil {
		*result = append(*result, node.Value)
		bt.preorder(node.Left, result)
		bt.preorder(node.Right, result)
	}
}

func (bt *BinaryTree) PostorderTraversal() []int {
	var result []int
	bt.postorder(bt.Root, &result)
	return result
}

func (bt *BinaryTree) postorder(node *TreeNode, result *[]int) {
	if node != nil {
		bt.postorder(node.Left, result)
		bt.postorder(node.Right, result)
		*result = append(*result, node.Value)
	}
}

// HashTable implementation
type HashTable struct {
	buckets [][]KeyValue
	size    int
}

type KeyValue struct {
	Key   string
	Value interface{}
}

func NewHashTable(size int) *HashTable {
	return &HashTable{
		buckets: make([][]KeyValue, size),
		size:    size,
	}
}

func (ht *HashTable) hash(key string) int {
	hash := 0
	for _, char := range key {
		hash += int(char)
	}
	return hash % ht.size
}

func (ht *HashTable) Set(key string, value interface{}) {
	index := ht.hash(key)
	
	for i, kv := range ht.buckets[index] {
		if kv.Key == key {
			ht.buckets[index][i].Value = value
			return
		}
	}
	
	ht.buckets[index] = append(ht.buckets[index], KeyValue{Key: key, Value: value})
}

func (ht *HashTable) Get(key string) (interface{}, error) {
	index := ht.hash(key)
	
	for _, kv := range ht.buckets[index] {
		if kv.Key == key {
			return kv.Value, nil
		}
	}
	
	return nil, fmt.Errorf("key '%s' not found", key)
}

func (ht *HashTable) Delete(key string) error {
	index := ht.hash(key)
	
	for i, kv := range ht.buckets[index] {
		if kv.Key == key {
			ht.buckets[index] = append(ht.buckets[index][:i], ht.buckets[index][i+1:]...)
			return nil
		}
	}
	
	return fmt.Errorf("key '%s' not found", key)
}

func (ht *HashTable) Keys() []string {
	var keys []string
	for _, bucket := range ht.buckets {
		for _, kv := range bucket {
			keys = append(keys, kv.Key)
		}
	}
	return keys
}

// Sorting algorithms
func BubbleSort(arr []int) []int {
	result := make([]int, len(arr))
	copy(result, arr)
	
	n := len(result)
	for i := 0; i < n-1; i++ {
		for j := 0; j < n-i-1; j++ {
			if result[j] > result[j+1] {
				result[j], result[j+1] = result[j+1], result[j]
			}
		}
	}
	return result
}

func QuickSort(arr []int) []int {
	result := make([]int, len(arr))
	copy(result, arr)
	quickSortHelper(result, 0, len(result)-1)
	return result
}

func quickSortHelper(arr []int, low, high int) {
	if low < high {
		pi := partition(arr, low, high)
		quickSortHelper(arr, low, pi-1)
		quickSortHelper(arr, pi+1, high)
	}
}

func partition(arr []int, low, high int) int {
	pivot := arr[high]
	i := low - 1
	
	for j := low; j < high; j++ {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i + 1
}

func MergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}
	
	mid := len(arr) / 2
	left := MergeSort(arr[:mid])
	right := MergeSort(arr[mid:])
	
	return merge(left, right)
}

func merge(left, right []int) []int {
	result := make([]int, 0, len(left)+len(right))
	i, j := 0, 0
	
	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}
	
	for i < len(left) {
		result = append(result, left[i])
		i++
	}
	
	for j < len(right) {
		result = append(result, right[j])
		j++
	}
	
	return result
}

// Mathematical utilities
func Factorial(n int) int {
	if n <= 1 {
		return 1
	}
	return n * Factorial(n-1)
}

func Fibonacci(n int) int {
	if n <= 1 {
		return n
	}
	return Fibonacci(n-1) + Fibonacci(n-2)
}

func IsPrime(n int) bool {
	if n <= 1 {
		return false
	}
	if n <= 3 {
		return true
	}
	if n%2 == 0 || n%3 == 0 {
		return false
	}
	
	for i := 5; i*i <= n; i += 6 {
		if n%i == 0 || n%(i+2) == 0 {
			return false
		}
	}
	return true
}

func GCD(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func LCM(a, b int) int {
	return (a * b) / GCD(a, b)
}

func main() {
	fmt.Println("Advanced Data Structures and Algorithms Demo")
	fmt.Println("============================================")
	
	// Stack demo
	fmt.Println("\n1. Stack Operations:")
	stack := NewStack()
	stack.Push(10)
	stack.Push(20)
	stack.Push(30)
	fmt.Printf("Stack size: %d\n", stack.Size())
	
	top, _ := stack.Peek()
	fmt.Printf("Top element: %v\n", top)
	
	popped, _ := stack.Pop()
	fmt.Printf("Popped: %v\n", popped)
	fmt.Printf("Stack size after pop: %d\n", stack.Size())
	
	// Queue demo
	fmt.Println("\n2. Queue Operations:")
	queue := NewQueue()
	queue.Enqueue("A")
	queue.Enqueue("B")
	queue.Enqueue("C")
	fmt.Printf("Queue size: %d\n", queue.Size())
	
	front, _ := queue.Front()
	fmt.Printf("Front element: %v\n", front)
	
	dequeued, _ := queue.Dequeue()
	fmt.Printf("Dequeued: %v\n", dequeued)
	fmt.Printf("Queue size after dequeue: %d\n", queue.Size())
	
	// Binary Tree demo
	fmt.Println("\n3. Binary Tree Operations:")
	tree := NewBinaryTree()
	values := []int{50, 30, 70, 20, 40, 60, 80}
	for _, v := range values {
		tree.Insert(v)
	}
	
	fmt.Printf("Inorder traversal: %v\n", tree.InorderTraversal())
	fmt.Printf("Preorder traversal: %v\n", tree.PreorderTraversal())
	fmt.Printf("Postorder traversal: %v\n", tree.PostorderTraversal())
	
	fmt.Printf("Search 40: %t\n", tree.Search(40))
	fmt.Printf("Search 90: %t\n", tree.Search(90))
	
	// Hash Table demo
	fmt.Println("\n4. Hash Table Operations:")
	ht := NewHashTable(10)
	ht.Set("name", "John")
	ht.Set("age", 25)
	ht.Set("city", "New York")
	
	name, _ := ht.Get("name")
	fmt.Printf("Name: %v\n", name)
	
	age, _ := ht.Get("age")
	fmt.Printf("Age: %v\n", age)
	
	fmt.Printf("All keys: %v\n", ht.Keys())
	
	// Sorting algorithms demo
	fmt.Println("\n5. Sorting Algorithms:")
	unsorted := []int{64, 34, 25, 12, 22, 11, 90}
	fmt.Printf("Unsorted array: %v\n", unsorted)
	
	bubbleSorted := BubbleSort(unsorted)
	fmt.Printf("Bubble sort: %v\n", bubbleSorted)
	
	quickSorted := QuickSort(unsorted)
	fmt.Printf("Quick sort: %v\n", quickSorted)
	
	mergeSorted := MergeSort(unsorted)
	fmt.Printf("Merge sort: %v\n", mergeSorted)
	
	// Mathematical functions demo
	fmt.Println("\n6. Mathematical Functions:")
	fmt.Printf("Factorial of 5: %d\n", Factorial(5))
	fmt.Printf("Fibonacci of 10: %d\n", Fibonacci(10))
	fmt.Printf("Is 17 prime: %t\n", IsPrime(17))
	fmt.Printf("Is 18 prime: %t\n", IsPrime(18))
	fmt.Printf("GCD of 48 and 18: %d\n", GCD(48, 18))
	fmt.Printf("LCM of 12 and 8: %d\n", LCM(12, 8))
	
	// Performance comparison
	fmt.Println("\n7. Performance Test:")
	largeArray := make([]int, 1000)
	for i := 0; i < 1000; i++ {
		largeArray[i] = 1000 - i
	}
	
	fmt.Printf("Sorting array of size %d\n", len(largeArray))
	
	// Test different sorting algorithms (commented out for brevity)
	// You would measure time here in a real scenario
	sorted1 := BubbleSort(largeArray[:100]) // Only first 100 elements for bubble sort
	fmt.Printf("Bubble sort (100 elements): %d...%d\n", sorted1[0], sorted1[len(sorted1)-1])
	
	sorted2 := QuickSort(largeArray)
	fmt.Printf("Quick sort (1000 elements): %d...%d\n", sorted2[0], sorted2[len(sorted2)-1])
	
	sorted3 := MergeSort(largeArray)
	fmt.Printf("Merge sort (1000 elements): %d...%d\n", sorted3[0], sorted3[len(sorted3)-1])
}

File: examples/python/data_structures.py
#!/usr/bin/env python3
"""
Data Structure Implementations and Algorithms
"""

from typing import List, Optional, Any, Iterator
import heapq
from collections import deque


class Node:
    """Basic node structure for linked lists and trees."""
    
    def __init__(self, data: Any):
        self.data = data
        self.next: Optional['Node'] = None
    
    def __str__(self) -> str:
        return str(self.data)


class LinkedList:
    """Implementation of a singly linked list."""
    
    def __init__(self):
        self.head: Optional[Node] = None
        self.size = 0
    
    def append(self, data: Any) -> None:
        """Add element to the end of the list."""
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
        self.size += 1
    
    def prepend(self, data: Any) -> None:
        """Add element to the beginning of the list."""
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
        self.size += 1
    
    def delete(self, data: Any) -> bool:
        """Delete first occurrence of data."""
        if not self.head:
            return False
        
        if self.head.data == data:
            self.head = self.head.next
            self.size -= 1
            return True
        
        current = self.head
        while current.next:
            if current.next.data == data:
                current.next = current.next.next
                self.size -= 1
                return True
            current = current.next
        
        return False
    
    def find(self, data: Any) -> bool:
        """Check if data exists in the list."""
        current = self.head
        while current:
            if current.data == data:
                return True
            current = current.next
        return False
    
    def get_at_index(self, index: int) -> Any:
        """Get element at specific index."""
        if index < 0 or index >= self.size:
            raise IndexError("Index out of range")
        
        current = self.head
        for _ in range(index):
            current = current.next
        return current.data
    
    def reverse(self) -> None:
        """Reverse the linked list in place."""
        prev = None
        current = self.head
        
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        
        self.head = prev
    
    def to_list(self) -> List[Any]:
        """Convert to Python list."""
        result = []
        current = self.head
        while current:
            result.append(current.data)
            current = current.next
        return result
    
    def __len__(self) -> int:
        return self.size
    
    def __str__(self) -> str:
        return " -> ".join(str(data) for data in self.to_list())


class Stack:
    """Implementation of a stack using a list."""
    
    def __init__(self):
        self.items: List[Any] = []
    
    def push(self, item: Any) -> None:
        """Add item to top of stack."""
        self.items.append(item)
    
    def pop(self) -> Any:
        """Remove and return top item."""
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self.items.pop()
    
    def peek(self) -> Any:
        """Return top item without removing it."""
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self.items[-1]
    
    def is_empty(self) -> bool:
        """Check if stack is empty."""
        return len(self.items) == 0
    
    def size(self) -> int:
        """Return size of stack."""
        return len(self.items)
    
    def __str__(self) -> str:
        return f"Stack({self.items})"


class Queue:
    """Implementation of a queue using deque."""
    
    def __init__(self):
        self.items = deque()
    
    def enqueue(self, item: Any) -> None:
        """Add item to rear of queue."""
        self.items.append(item)
    
    def dequeue(self) -> Any:
        """Remove and return front item."""
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items.popleft()
    
    def front(self) -> Any:
        """Return front item without removing it."""
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items[0]
    
    def is_empty(self) -> bool:
        """Check if queue is empty."""
        return len(self.items) == 0
    
    def size(self) -> int:
        """Return size of queue."""
        return len(self.items)
    
    def __str__(self) -> str:
        return f"Queue({list(self.items)})"


class MinHeap:
    """Implementation of a min heap."""
    
    def __init__(self):
        self.heap: List[Any] = []
    
    def push(self, item: Any) -> None:
        """Add item to heap."""
        heapq.heappush(self.heap, item)
    
    def pop(self) -> Any:
        """Remove and return minimum item."""
        if self.is_empty():
            raise IndexError("Heap is empty")
        return heapq.heappop(self.heap)
    
    def peek(self) -> Any:
        """Return minimum item without removing it."""
        if self.is_empty():
            raise IndexError("Heap is empty")
        return self.heap[0]
    
    def is_empty(self) -> bool:
        """Check if heap is empty."""
        return len(self.heap) == 0
    
    def size(self) -> int:
        """Return size of heap."""
        return len(self.heap)
    
    def __str__(self) -> str:
        return f"MinHeap({self.heap})"


class SortingAlgorithms:
    """Collection of sorting algorithms."""
    
    @staticmethod
    def bubble_sort(arr: List[Any]) -> List[Any]:
        """Bubble sort implementation."""
        arr = arr.copy()
        n = len(arr)
        
        for i in range(n):
            swapped = False
            for j in range(0, n - i - 1):
                if arr[j] > arr[j + 1]:
                    arr[j], arr[j + 1] = arr[j + 1], arr[j]
                    swapped = True
            if not swapped:
                break
        
        return arr
    
    @staticmethod
    def selection_sort(arr: List[Any]) -> List[Any]:
        """Selection sort implementation."""
        arr = arr.copy()
        n = len(arr)
        
        for i in range(n):
            min_idx = i
            for j in range(i + 1, n):
                if arr[j] < arr[min_idx]:
                    min_idx = j
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
        
        return arr
    
    @staticmethod
    def insertion_sort(arr: List[Any]) -> List[Any]:
        """Insertion sort implementation."""
        arr = arr.copy()
        
        for i in range(1, len(arr)):
            key = arr[i]
            j = i - 1
            while j >= 0 and arr[j] > key:
                arr[j + 1] = arr[j]
                j -= 1
            arr[j + 1] = key
        
        return arr
    
    @staticmethod
    def merge_sort(arr: List[Any]) -> List[Any]:
        """Merge sort implementation."""
        if len(arr) <= 1:
            return arr.copy()
        
        mid = len(arr) // 2
        left = SortingAlgorithms.merge_sort(arr[:mid])
        right = SortingAlgorithms.merge_sort(arr[mid:])
        
        return SortingAlgorithms._merge(left, right)
    
    @staticmethod
    def _merge(left: List[Any], right: List[Any]) -> List[Any]:
        """Helper function for merge sort."""
        result = []
        i = j = 0
        
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1
        
        result.extend(left[i:])
        result.extend(right[j:])
        return result
    
    @staticmethod
    def quick_sort(arr: List[Any]) -> List[Any]:
        """Quick sort implementation."""
        if len(arr) <= 1:
            return arr.copy()
        
        pivot = arr[len(arr) // 2]
        left = [x for x in arr if x < pivot]
        middle = [x for x in arr if x == pivot]
        right = [x for x in arr if x > pivot]
        
        return (SortingAlgorithms.quick_sort(left) + 
                middle + 
                SortingAlgorithms.quick_sort(right))


class SearchAlgorithms:
    """Collection of search algorithms."""
    
    @staticmethod
    def linear_search(arr: List[Any], target: Any) -> int:
        """Linear search implementation."""
        for i, item in enumerate(arr):
            if item == target:
                return i
        return -1
    
    @staticmethod
    def binary_search(arr: List[Any], target: Any) -> int:
        """Binary search implementation (requires sorted array)."""
        left, right = 0, len(arr) - 1
        
        while left <= right:
            mid = (left + right) // 2
            if arr[mid] == target:
                return mid
            elif arr[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        
        return -1


def demonstrate_data_structures():
    """Demonstrate data structure implementations."""
    print("=== Data Structures and Algorithms Demo ===")
    
    # Linked List
    print("\n--- Linked List ---")
    ll = LinkedList()
    for i in range(1, 6):
        ll.append(i)
    print(f"Original list: {ll}")
    ll.prepend(0)
    print(f"After prepend(0): {ll}")
    ll.delete(3)
    print(f"After delete(3): {ll}")
    ll.reverse()
    print(f"After reverse: {ll}")
    print(f"Element at index 2: {ll.get_at_index(2)}")
    print(f"Contains 4: {ll.find(4)}")
    
    # Stack
    print("\n--- Stack ---")
    stack = Stack()
    for i in range(1, 4):
        stack.push(i)
        print(f"Pushed {i}: {stack}")
    
    while not stack.is_empty():
        popped = stack.pop()
        print(f"Popped {popped}: {stack}")
    
    # Queue
    print("\n--- Queue ---")
    queue = Queue()
    for i in range(1, 4):
        queue.enqueue(i)
        print(f"Enqueued {i}: {queue}")
    
    while not queue.is_empty():
        dequeued = queue.dequeue()
        print(f"Dequeued {dequeued}: {queue}")
    
    # Min Heap
    print("\n--- Min Heap ---")
    heap = MinHeap()
    values = [5, 2, 8, 1, 9, 3]
    for val in values:
        heap.push(val)
        print(f"Pushed {val}: {heap}")
    
    print("Popping all elements:")
    while not heap.is_empty():
        min_val = heap.pop()
        print(f"Popped {min_val}: {heap}")
    
    # Sorting Algorithms
    print("\n--- Sorting Algorithms ---")
    test_array = [64, 34, 25, 12, 22, 11, 90]
    print(f"Original array: {test_array}")
    
    sorters = SortingAlgorithms()
    print(f"Bubble sort: {sorters.bubble_sort(test_array)}")
    print(f"Selection sort: {sorters.selection_sort(test_array)}")
    print(f"Insertion sort: {sorters.insertion_sort(test_array)}")
    print(f"Merge sort: {sorters.merge_sort(test_array)}")
    print(f"Quick sort: {sorters.quick_sort(test_array)}")
    
    # Search Algorithms
    print("\n--- Search Algorithms ---")
    sorted_array = sorted(test_array)
    print(f"Sorted array: {sorted_array}")
    
    target = 25
    linear_result = SearchAlgorithms.linear_search(test_array, target)
    binary_result = SearchAlgorithms.binary_search(sorted_array, target)
    
    print(f"Linear search for {target}: index {linear_result}")
    print(f"Binary search for {target}: index {binary_result}")
    
    # Performance comparison
    import time
    large_array = list(range(10000, 0, -1))  # Reverse sorted array
    
    print("\n--- Performance Test (10,000 elements) ---")
    algorithms = [
        ("Bubble Sort", sorters.bubble_sort),
        ("Selection Sort", sorters.selection_sort),
        ("Insertion Sort", sorters.insertion_sort),
        ("Merge Sort", sorters.merge_sort),
        ("Quick Sort", sorters.quick_sort),
    ]
    
    test_size = 1000  # Smaller test for bubble/selection sort
    small_array = large_array[:test_size]
    
    for name, algorithm in algorithms:
        start_time = time.time()
        if name in ["Bubble Sort", "Selection Sort"]:
            # Test with smaller array for slower algorithms
            algorithm(small_array)
            array_size = test_size
        else:
            algorithm(large_array)
            array_size = len(large_array)
        end_time = time.time()
        
        print(f"{name} ({array_size} elements): {end_time - start_time:.4f} seconds")


if __name__ == "__main__":
    demonstrate_data_structures()